api/core/features/assistant_fc_runner.py

```diff
@@ -0,0 +1,335 @@
+                llm_usage.prompt_price += usage.prompt_price
+                llm_usage.completion_price += usage.completion_price
+
+        while function_call_state and iteration_step <= max_iteration_steps:
+            function_call_state = False
+
+            if iteration_step == max_iteration_steps:
+                # the last iteration, remove all tools
+                prompt_messages_tools = []
+
+            message_file_ids = []
+            agent_thought = self.create_agent_thought(
+                message_id=message.id,
+                message='',
+                tool_name='',
+                tool_input='',
+                messages_ids=message_file_ids
+            )
+            self.queue_manager.publish_agent_thought(agent_thought, PublishFrom.APPLICATION_MANAGER)
+
+            # recale llm max tokens
+            self.recale_llm_max_tokens(self.model_config, prompt_messages)
+            # invoke model
+            chunks: Generator[LLMResultChunk, None, None] = model_instance.invoke_llm(
+                prompt_messages=prompt_messages,
+                model_parameters=app_orchestration_config.model_config.parameters,
+                tools=prompt_messages_tools,
+                stop=app_orchestration_config.model_config.stop,
+                stream=True,
+                user=self.user_id,
+                callbacks=[],
+            )
+
+            tool_calls: List[Tuple[str, str, Dict[str, Any]]] = []
+
+            # save full response
+            response = ''
+
+            # save tool call names and inputs
+            tool_call_names = ''
+            tool_call_inputs = ''
+
+            current_llm_usage = None
+
+            for chunk in chunks:
+                # check if there is any tool call
+                if self.check_tool_calls(chunk):
+                    function_call_state = True
+                    tool_calls.extend(self.extract_tool_calls(chunk))
+                    tool_call_names = ';'.join([tool_call[1] for tool_call in tool_calls])
+                    try:
+                        tool_call_inputs = json.dumps({
+                            tool_call[1]: tool_call[2] for tool_call in tool_calls
+                        }, ensure_ascii=False)
+                    except json.JSONDecodeError as e:
+                        # ensure ascii to avoid encoding error
+                        tool_call_inputs = json.dumps({
+                            tool_call[1]: tool_call[2] for tool_call in tool_calls
+                        })
+
+                if chunk.delta.message and chunk.delta.message.content:
+                    if isinstance(chunk.delta.message.content, list):
+                        for content in chunk.delta.message.content:
+                            response += content.data
+                    else:
+                        response += chunk.delta.message.content
+
+                if chunk.delta.usage:
+                    increse_usage(llm_usage, chunk.delta.usage)
```

Comment:
Did you mean increase?
```suggestion
                    increase_usage(llm_usage, chunk.delta.usage)
```